File: problems/59/water_container.c
#include <criterion/criterion.h>
#include <l33t.h>

int max_area(int *heights, int num)
{
  if (heights == NULL || num <= 0)
  {
    return 0;
  }

  int max = 0;
  int left = 0;
  int right = num - 1;
  while (left < right)
  {
    const int area = MIN(heights[left], heights[right]) * (right - left);
    max = area > max ? area : max;
    heights[left] < heights[right] ? left++ : right--;
  }

  return max;
}

TestSuite(WaterContainerTests);
Test(WaterContainerTests, LeetcodeExamples)
{
  typedef struct
  {
    int heights[9];
    int num;
    int expected;
  } input;
  input inputs[5] = {
      {.heights = {1, 8, 6, 2, 5, 4, 8, 3, 7}, .num = 9, .expected = 49},
      {.heights = {1, 1}, .num = 2, .expected = 1},
      {.heights = {1, 2, 4, 3}, .num = 4, .expected = 4},
      {.heights = {2, 3, 4, 5, 18, 17, 6}, .num = 7, .expected = 17},
      {.heights = {1, 0, 0, 0, 0, 0, 0, 2, 2}, .num = 9, .expected = 8},
  };
  for (int i = 0; i < (int)ARRAY_SIZE(inputs, input); i++)
  {
    input in = inputs[i];
    const int area = max_area(in.heights, in.num);
    cr_expect(area == in.expected);
  }
}

File: problems/59/water_container.py
def get_area(h, l, r):
    return min(h[l], h[r]) * (r - l)


def max_area(heights: list[int]) -> int:
    maximum = 0
    left = 0
    right = len(heights) - 1
    while left < right:
        area = get_area(heights, left, right)
        if area > maximum:
            maximum = area

        if heights[left] < heights[right]:
            left += 1
        else:
            right -= 1

    return maximum


def test_water_container():
    assert max_area([1, 8, 6, 2, 5, 4, 8, 3, 7]) == 49
    assert max_area([1, 1]) == 1
    assert max_area([1, 2, 4, 3]) == 4
    assert max_area([2, 3, 4, 5, 18, 17, 6]) == 17
    assert max_area([1, 0, 0, 0, 0, 0, 0, 2, 2]) == 8


if __name__ == "__main__":
    import pytest

    pytest.main(["-xvv", __file__])
